File: typescript/QueryEngine.ts
// =============================================================================
// vapor-idx — QueryEngine.ts
// Executes queries against RecordStore indexes.
// Handles type filtering, field filters (by index strategy), keyword search,
// AND/OR logic, ordering, limit, and offset.
// =============================================================================

import {
  VaporRecord,
  QueryOptions,
  QueryResult,
  FieldFilter,
  FilterOp,
  VaporQueryError,
} from './types.js';
import { RecordStore }   from './RecordStore.js';
import { validateFilter } from './SchemaValidator.js';

// Ops that read from the range index for numeric ordering support
const RANGE_OPS = new Set<FilterOp>(['gt', 'lt', 'gte', 'lte']);

export class QueryEngine {
  constructor(private readonly store: RecordStore) {}

  query<T extends Record<string, unknown>>(options: QueryOptions): QueryResult<T> {
    const types = resolveTypes(options.type, this.store.getTypes());

    // 1. Start with candidate ID sets per type, then union
    let candidateIds: Set<string> | null = null;

    for (const typeName of types) {
      const typeDef   = this.store['schema'].types[typeName];
      const typeIdSet = new Set(this.store.getTypeIdSet(typeName));

      if (typeIdSet.size === 0) continue;

      // Apply where filters for this type
      const filters = normaliseFilters(options.where);

      // Validate filters against the type
      for (const filter of filters) {
        validateFilter(filter, typeName, typeDef);
      }

      const filtered = this.applyFilters(typeName, typeIdSet, filters, options.logic ?? 'AND');

      // Apply keywords
      const afterKeywords = this.applyKeywords(typeName, filtered, options.keywords);

      // Merge across types
      if (candidateIds === null) {
        candidateIds = afterKeywords;
      } else {
        for (const id of afterKeywords) candidateIds.add(id);
      }
    }

    if (candidateIds === null || candidateIds.size === 0) {
      return { records: [], total: 0 };
    }

    // 2. Hydrate records
    let records = [...candidateIds]
      .map(id => this.store.get(id))
      .filter((r): r is VaporRecord<T> => r !== null) as VaporRecord<T>[];

    // 3. Apply orderBy
    if (options.orderBy) {
      records = sortRecords(records, options.orderBy.field, options.orderBy.direction);
    }

    const total = records.length;

    // 4. Apply offset and limit
    const offset = options.offset ?? 0;
    const limit  = options.limit;

    records = records.slice(offset, limit !== undefined ? offset + limit : undefined);

    return { records, total };
  }

  // ── Filter application ─────────────────────────────────────────────────────

  private applyFilters(
    typeName:  string,
    seed:      Set<string>,
    filters:   FieldFilter[],
    logic:     'AND' | 'OR'
  ): Set<string> {
    if (filters.length === 0) return seed;

    const ti = this.store.getIndexes(typeName)!;

    if (logic === 'AND') {
      let current = seed;
      for (const filter of filters) {
        const matched = this.applyFilter(filter, typeName);
        current       = intersection(current, matched);
        if (current.size === 0) break;
      }
      return current;
    } else {
      // OR: union results of each filter, then intersect with seed
      const unionResult = new Set<string>();
      for (const filter of filters) {
        const matched = this.applyFilter(filter, typeName);
        for (const id of matched) {
          if (seed.has(id)) unionResult.add(id);
        }
      }
      return unionResult;
    }
  }

  private applyFilter(filter: FieldFilter, typeName: string): Set<string> {
    const ti = this.store.getIndexes(typeName)!;
    const { field, op, value } = filter;

    switch (op as FilterOp) {
      case 'eq':         return ti.exact.eq(field, value);
      case 'neq':        return ti.exact.neq(field, value);
      case 'in':         return ti.exact.findIn(field, value as unknown[]);
      case 'notIn':      return ti.exact.notIn(field, value as unknown[]);
      case 'contains':   return ti.keyword.contains(field, value as string);
      case 'startsWith': return ti.prefix.startsWith(field, value as string);
      case 'gt':         return ti.range.gt(field, value as number);
      case 'lt':         return ti.range.lt(field, value as number);
      case 'gte':        return ti.range.gte(field, value as number);
      case 'lte':        return ti.range.lte(field, value as number);
      default:
        throw new VaporQueryError(`Unknown filter operator "${op}".`);
    }
  }

  private applyKeywords(
    typeName: string,
    seed:     Set<string>,
    keywords: string | string[] | undefined
  ): Set<string> {
    if (!keywords || (Array.isArray(keywords) && keywords.length === 0)) return seed;

    const ti      = this.store.getIndexes(typeName)!;
    const matched = ti.keyword.search(keywords);
    return intersection(seed, matched);
  }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function resolveTypes(
  typeOption: string | string[] | undefined,
  allTypes:   string[]
): string[] {
  if (typeOption === undefined) return allTypes;
  if (Array.isArray(typeOption)) return typeOption;
  return [typeOption];
}

function normaliseFilters(where: QueryOptions['where']): FieldFilter[] {
  if (!where) return [];
  if (Array.isArray(where)) return where;
  return [where];
}

function intersection(a: Set<string>, b: Set<string>): Set<string> {
  const [small, large] = a.size <= b.size ? [a, b] : [b, a];
  const result         = new Set<string>();
  for (const id of small) {
    if (large.has(id)) result.add(id);
  }
  return result;
}

function sortRecords<T extends Record<string, unknown>>(
  records:   VaporRecord<T>[],
  field:     string,
  direction: 'asc' | 'desc'
): VaporRecord<T>[] {
  return [...records].sort((a, b) => {
    const av = a.data[field] as string | number | boolean | null | undefined;
    const bv = b.data[field] as string | number | boolean | null | undefined;

    if (av === null || av === undefined) return direction === 'asc' ? 1 : -1;
    if (bv === null || bv === undefined) return direction === 'asc' ? -1 : 1;

    if (typeof av === 'number' && typeof bv === 'number') {
      return direction === 'asc' ? av - bv : bv - av;
    }

    const as = String(av).toLowerCase();
    const bs = String(bv).toLowerCase();
    const cmp = as < bs ? -1 : as > bs ? 1 : 0;
    return direction === 'asc' ? cmp : -cmp;
  });
}

File: rust/src/query_engine.rs
// =============================================================================
// vapor-idx — query_engine.rs
// Imports validate_filter from schema.rs — no local re-definition.
// =============================================================================

use crate::record_store::RecordStore;
use crate::schema::validate_filter;
use crate::types::*;
use std::collections::HashSet;

pub struct QueryEngine<'a> {
    store: &'a RecordStore,
}

impl<'a> QueryEngine<'a> {
    pub fn new(store: &'a RecordStore) -> Self {
        Self { store }
    }

    pub fn query(&self, options: &QueryOptions) -> VaporResult<QueryResult> {
        let types = match &options.type_filter {
            Some(t) => t.clone(),
            None => self.store.get_types(),
        };
        let mut candidate_ids: Option<HashSet<String>> = None;

        for type_name in &types {
            let type_def = match self.store.schema.types.get(type_name) {
                Some(t) => t,
                None => continue,
            };
            let type_id_set = self.store.get_type_id_set(type_name);
            if type_id_set.is_empty() {
                continue;
            }

            for filter in &options.where_filters {
                validate_filter(filter, type_name, type_def)?;
            }

            let filtered = self.apply_filters(
                type_name,
                type_id_set,
                &options.where_filters,
                &options.logic,
            );
            let after_keywords = self.apply_keywords(type_name, filtered, &options.keywords);

            match &mut candidate_ids {
                None => {
                    candidate_ids = Some(after_keywords);
                }
                Some(ids) => {
                    ids.extend(after_keywords);
                }
            }
        }

        let ids = match candidate_ids {
            None => {
                return Ok(QueryResult {
                    records: vec![],
                    total: 0,
                })
            }
            Some(s) => s,
        };

        let mut records: Vec<VaporRecord> = ids
            .iter()
            .filter_map(|id| self.store.get(id))
            .cloned()
            .collect();

        if let Some((field, dir)) = &options.order_by {
            records.sort_by(|a, b| {
                let av = a.data.get(field);
                let bv = b.data.get(field);
                let cmp = compare_json(av, bv);
                if *dir == SortDirection::Desc {
                    cmp.reverse()
                } else {
                    cmp
                }
            });
        }

        let total = records.len();
        let offset = options.offset;
        let records = if let Some(limit) = options.limit {
            records.into_iter().skip(offset).take(limit).collect()
        } else {
            records.into_iter().skip(offset).collect()
        };

        Ok(QueryResult { records, total })
    }

    fn apply_filters(
        &self,
        type_name: &str,
        seed: HashSet<String>,
        filters: &[FieldFilter],
        logic: &QueryLogic,
    ) -> HashSet<String> {
        if filters.is_empty() {
            return seed;
        }
        match logic {
            QueryLogic::And => {
                let mut current = seed;
                for filter in filters {
                    let matched = self.apply_filter(type_name, filter);
                    current = current.intersection(&matched).cloned().collect();
                    if current.is_empty() {
                        break;
                    }
                }
                current
            }
            QueryLogic::Or => {
                let mut union_result = HashSet::new();
                for filter in filters {
                    let matched = self.apply_filter(type_name, filter);
                    let intersected: HashSet<String> =
                        matched.intersection(&seed).cloned().collect();
                    union_result.extend(intersected);
                }
                union_result
            }
        }
    }

    fn apply_filter(&self, type_name: &str, filter: &FieldFilter) -> HashSet<String> {
        let field = &filter.field;
        let value = &filter.value;
        match &filter.op {
            FilterOp::Eq => self
                .store
                .exact(type_name)
                .map(|i| i.eq(field, value))
                .unwrap_or_default(),
            FilterOp::Neq => self
                .store
                .exact(type_name)
                .map(|i| i.neq(field, value))
                .unwrap_or_default(),
            FilterOp::In => {
                let v = value.as_array().cloned().unwrap_or_default();
                self.store
                    .exact(type_name)
                    .map(|i| i.find_in(field, &v))
                    .unwrap_or_default()
            }
            FilterOp::NotIn => {
                let v = value.as_array().cloned().unwrap_or_default();
                self.store
                    .exact(type_name)
                    .map(|i| i.not_in(field, &v))
                    .unwrap_or_default()
            }
            FilterOp::Contains => {
                let q = value
                    .as_str()
                    .map(|s| vec![s.to_owned()])
                    .unwrap_or_default();
                self.store
                    .keyword(type_name)
                    .map(|i| i.contains(field, &q))
                    .unwrap_or_default()
            }
            FilterOp::StartsWith => {
                let p = value.as_str().unwrap_or("");
                self.store
                    .prefix(type_name)
                    .map(|i| i.starts_with(field, p))
                    .unwrap_or_default()
            }
            FilterOp::Gt => {
                let n = value.as_f64().unwrap_or(0.0);
                self.store
                    .range(type_name)
                    .map(|i| i.gt(field, n))
                    .unwrap_or_default()
            }
            FilterOp::Lt => {
                let n = value.as_f64().unwrap_or(0.0);
                self.store
                    .range(type_name)
                    .map(|i| i.lt(field, n))
                    .unwrap_or_default()
            }
            FilterOp::Gte => {
                let n = value.as_f64().unwrap_or(0.0);
                self.store
                    .range(type_name)
                    .map(|i| i.gte(field, n))
                    .unwrap_or_default()
            }
            FilterOp::Lte => {
                let n = value.as_f64().unwrap_or(0.0);
                self.store
                    .range(type_name)
                    .map(|i| i.lte(field, n))
                    .unwrap_or_default()
            }
        }
    }

    fn apply_keywords(
        &self,
        type_name: &str,
        seed: HashSet<String>,
        keywords: &Option<Vec<String>>,
    ) -> HashSet<String> {
        let Some(kws) = keywords else { return seed };
        if kws.is_empty() {
            return seed;
        }
        let matched = self
            .store
            .keyword(type_name)
            .map(|i| i.search(kws))
            .unwrap_or_default();
        seed.intersection(&matched).cloned().collect()
    }
}

fn compare_json(
    a: Option<&serde_json::Value>,
    b: Option<&serde_json::Value>,
) -> std::cmp::Ordering {
    use std::cmp::Ordering;
    match (a, b) {
        (None, None) => Ordering::Equal,
        (None, Some(_)) => Ordering::Greater,
        (Some(_), None) => Ordering::Less,
        (Some(a), Some(b)) => {
            if let (Some(an), Some(bn)) = (a.as_f64(), b.as_f64()) {
                an.partial_cmp(&bn).unwrap_or(Ordering::Equal)
            } else {
                a.as_str()
                    .unwrap_or("")
                    .to_lowercase()
                    .cmp(&b.as_str().unwrap_or("").to_lowercase())
            }
        }
    }
}
